File: PDF Decryptor/cpp/pdfdecryptor.cpp
/*
 * SecureBlackbox 2022 C++ Edition - Demo Application
 *
 * Copyright (c) 2023 /n software inc. - All rights reserved. - www.nsoftware.com
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../../include/secureblackbox.h"
#define LINE_LEN 100

namespace ArgParser {
    static char* optval(int argc, char** argv, const char* option) {
        for (int x = 0; x < argc - 1; x++) {
            if (!strcmp(argv[x], option)) {
                return argv[x + 1];
            }
        }
        return (char*)"";
    }

    static bool optext(int argc, char** argv, const char* option) {
        for (int x = 0; x < argc; x++) {
            if (!strcmp(argv[x], option)) {
                return true;
            }
        }
        return false;
    }
};

using namespace ArgParser;
PDFDecryptor pdfdecryptor;
CertificateManager certificatemanager;

int showHelp() {
    printf("Usage:\n"
        "\t./pdfdecryptor\t-inputFile <inputFile> -outputFile <outputFile>\n"
        "\t\t\t(-keyPath <path> -keyPassword <password> | -userPassword <password>)\n");
    printf("Options: \n");
    printf("\t-inputFile     (string) The PDF file to be decrypted.\n");
    printf("\t-outputFile    (string) Where to save the decrypted document.\n");
    printf("\t-keyPath       (string) The key file to be imported.\n");
    printf("\t-keyPassword   (string) The key password.\n");
    printf("\t-userPassword  (string) The password used for decryption.\n");
    exit(1);
}

int main(int argc, char** argv) {
    if (argc < 7) {
        fprintf(stderr, "Error: Missing arguments.\n");
        showHelp();
    }

    optext(argc, argv, "-inputFile") ? pdfdecryptor.SetInputFile(optval(argc, argv, "-inputFile")) : showHelp();
    optext(argc, argv, "-outputFile") ? pdfdecryptor.SetOutputFile(optval(argc, argv, "-outputFile")) : showHelp();

    if (optext(argc, argv, "-keyPath") && optext(argc, argv, "-keyPassword")) {
        char* keyPath = optval(argc, argv, "-keyPath");
        char* keyPassword = optval(argc, argv, "-keyPassword");
        certificatemanager.ImportFromFile(keyPath, keyPassword);
        pdfdecryptor.SetDecryptionCertCount(1);
        pdfdecryptor.SetDecryptionCertHandle(0, certificatemanager.GetCertHandle());
    } else if (optext(argc, argv, "-userPassword")) {
        char* password = optval(argc, argv, "-userPassword");
        pdfdecryptor.SetPassword(password);
    } else {
        showHelp();
    }
    const int retcode = pdfdecryptor.Decrypt();

    if (retcode) {
        printf("Error [%d]: %s", pdfdecryptor.GetLastErrorCode(), pdfdecryptor.GetLastError());
        return 0;
    } else {
        printf("The document was decrypted successfully.\n");
    };

    fprintf(stderr, "\npress <return> to continue...\n");
    getchar();
    return 0;

}

File: PDF Decryptor/python/pdf_decryptor.py
#
# SecureBlackbox 2022 Python Edition - Demo Application
#
# Copyright (c) 2023 /n software inc. - All rights reserved. - www.nsoftware.com
#

import sys
import string
from secureblackbox import *

input = sys.hexversion<0x03000000 and raw_input or input

def displayHelp():
    print(
        "NAME\n"
        "  pdfdecryptor -- SecureBlackbox PDFDecryptor Demo Application\n\n"
        "SYNOPSIS\n"
        "  pdfdecryptor [-input input_file] [-output output_file] [-cert certificate_file] [-certpass certificate_password] [-pass user_password] \n\n"
        "DESCRIPTION\n"
        "  PDFDecryptor demonstrates the usage of PDFDecryptor from SecureBlackbox.\n"
        "  Used to decrypt pdf documents.\n\n"
        "  The options are as follows:\n\n"
        "  -input        An input file to decrypt (Required).\n\n"
        "  -output       Where the decrypted file will be saved (Required).\n\n"
        "  -cert         The certificate used to decrypt file.\n\n"
        "  -certpass     The password for the certificate.\n\n"
        "  -pass         The password for decryption.\n\n"
        "EXAMPLES\n"
        "  pdfdecryptor -input C:\\pdf\\mypdf.scs -output C:\\pdf\\helloworld.pdf -cert C:\\certs\\mycert.pfx -certpass mypassword \n\n"
        "  pdfdecryptor -input C:\\pdf\\mypdf.scs -output C:\\pdf\\helloworld.pdf -pass mypassword \n\n"
    )
    
if (len(sys.argv) < 5):
    displayHelp()
    sys.exit(1)
else:
    decryptor = PDFDecryptor()
    cm = CertificateManager()
    
    inputF = ""
    outputF = ""
    certF = ""
    certpass = ""
    password = ""

    for x in range(len(sys.argv)):
        if (sys.argv[x].startswith("-")):
            if (sys.argv[x].lower() == "-input"):
                inputF = sys.argv[x+1]
            if (sys.argv[x].lower() == "-output"):
                outputF = sys.argv[x+1]
            if (sys.argv[x].lower() == "-cert"):
                certF = sys.argv[x+1]
            if (sys.argv[x].lower() == "-certpass"):
                certpass = sys.argv[x+1]
            if (sys.argv[x].lower() == "-pass"):
                password = sys.argv[x+1]
    
    if (inputF == ""):
        print("-input is required.\n")
        displayHelp()
        sys.exit(1)
    
    if (outputF == ""):
        print("-output is required.\n")
        displayHelp()
        sys.exit(1)
    
    if (certF != ""):
        cm.import_from_file(certF, certpass)
        decryptor.set_decryption_certificate_handle(cm.get_cert_handle())
    else:
        decryptor.set_password(password)

    try:
        decryptor.set_input_file(inputF)
        decryptor.set_output_file(outputF)

        decryptor.decrypt()
    
        print("PDF file successfully decrypted.")
    except Exception as e: 
        print(e)



